In [1]:
import pandas as pd

### 지도학습

#### 분류 - 이진분류
- 목표변수: 범주형 (0 or 1)
- 설명변수: 


In [2]:
df_TFD = pd.read_csv('../../../datasets/TitanicFromDisaster_train.csv')
df_TFD

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

#### data preprocessing _ 전처리
- 목표변수: Survived (범주형) 
- 설명변수: Pclass (범주형), Age(연속형)

In [3]:
df_TFD.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [4]:
df_TFD_extract = df_TFD[['Survived', 'Pclass','Age' ]]

In [5]:
df_TFD_extract.isnull().sum()

Survived      0
Pclass        0
Age         177
dtype: int64

In [41]:
df_TFD_extract_preprocess = df_TFD_extract.dropna()
df_TFD_extract_preprocess

Survived  Pclass   Age
0           0       3  22.0
1           1       1  38.0
2           1       3  26.0
3           1       1  35.0
4           0       3  35.0
..        ...     ...   ...
885         0       3  39.0
886         0       2  27.0
887         1       1  19.0
889         1       1  26.0
890         0       3  32.0

[714 rows x 3 columns]

#### Scaling & Encoding
- Encoding: 범주형일때 진행함
    - 일반적으로 train_test_split 하기 전에 진행한다
    

##### Encoding with OneHotEncoding


In [7]:
df_TFD_extract_preprocess['Pclass'].value_counts()

3    355
1    186
2    173
Name: Pclass, dtype: int64

In [8]:
from sklearn.preprocessing import OneHotEncoder

In [112]:
oneHotEncoder = OneHotEncoder()

In [113]:
oneHotEncoder.fit(df_TFD_extract_preprocess[['Pclass']]) ## onehotencoder로 범주형을 학습한것. 

OneHotEncoder()

In [114]:
columns_name = oneHotEncoder.categories_

In [115]:
# eHotEncoder.transform(df_TFD_extract_preprocess[['Pclass']]).toarray() # 1,2,3을 transform 시킴 _값 확인용
encoded_data = oneHotEncoder.transform(df_TFD_extract_preprocess[['Pclass']]).toarray()

In [118]:
df_encoed_data = pd.DataFrame(data=encoded_data, columns=columns_name)
df_encoed_data

1    2    3
0    0.0  0.0  1.0
1    1.0  0.0  0.0
2    0.0  0.0  1.0
3    1.0  0.0  0.0
4    0.0  0.0  1.0
..   ...  ...  ...
709  0.0  0.0  1.0
710  0.0  1.0  0.0
711  1.0  0.0  0.0
712  1.0  0.0  0.0
713  0.0  0.0  1.0

[714 rows x 3 columns]

In [119]:
df_encoed_data.index, df_encoded_data.shape

(RangeIndex(start=0, stop=714, step=1), (714, 3))

In [120]:
df_TFD_extract_preprocess.index, df_TFD_extract_preprocess.shape

(Int64Index([  0,   1,   2,   3,   4,   6,   7,   8,   9,  10,
             ...
             880, 881, 882, 883, 884, 885, 886, 887, 889, 890],
            dtype='int64', length=714),
 (714, 6))

In [108]:
# df_encoded_data=pd.get_dummies(df_TFD_extract_preprocess['Pclass'], prefix='Pclass') # 만든거 합친다. 붙일 두개 나열

In [79]:
df_TFD_extract_preprocess=pd.concat([df_TFD_extract_preprocess, df_encoded_data], axis=1) # index값이 일치해야함. 

In [80]:
df_TFD_extract_preprocess[['Pclass']] ## Pclass

Pclass
0         3
1         1
2         3
3         1
4         3
..      ...
885       3
886       2
887       1
889       1
890       3

[714 rows x 1 columns]

In [81]:
target = df_TFD_extract_preprocess['Survived']

In [82]:
features = df_TFD_extract_preprocess.drop(columns=['Survived', 'Pclass'])

In [83]:
features.columns

Index(['Age', 'Pclass_1', 'Pclass_2', 'Pclass_3'], dtype='object')

#### MinMaxScaler

In [84]:
from sklearn.preprocessing import MinMaxScaler

In [86]:
minMaxScaler=MinMaxScaler()

In [88]:
features=minMaxScaler.fit_transform(features)
features.shape

(714, 4)

#### structured data / 정형화 단계
- 7:3으로 data를 나눌거다.
   - from sklearn.model_selection import train_test_split
   - features_train, features_test, target_train, target_test = train_test_split(features, target, random_state =111 ← random_state 로 고정을 한다.  )
   - 500개 이하는 split NO NO NO. 

In [89]:
from sklearn.model_selection import train_test_split


In [90]:
features_train, features_test, target_train, target_test = train_test_split(features, target, random_state= 111)
features_train.shape, features_test.shape, target_train.shape, target_test.shape

((535, 4), (179, 4), (535,), (179,))

#### Learning _모델 학습

In [91]:
from sklearn.linear_model import LogisticRegression

In [92]:
model = LogisticRegression()

In [93]:
model.fit(features_train,target_train)

LogisticRegression()

###### Logistic Regression 
y = 1/(1+e^-(wx+b))

In [96]:
model.coef_, model.intercept_
#이 공식은 logistics Regression 의 공식의 각 변수들의 상수임. 
# 정규화 이전값: (array([-1.25703902, -0.04521969]), array([3.78621796]))

(array([[-2.59255724,  1.11141847,  0.02997489, -1.14146766]]),
 array([0.86447724]))

#### 예측

In [97]:
df_TFD_extract_preprocess[250:255] # index가 features_train 순서와 동일함. 

Survived  Pclass   Age  Pclass_1  Pclass_2  Pclass_3
313         0       3  28.0         0         0         1
314         0       2  43.0         0         1         0
315         1       3  26.0         0         0         1
316         1       2  24.0         0         1         0
317         0       2  54.0         0         1         0

In [98]:
model.predict(features_train[250:255])

array([1, 1, 1, 1, 0], dtype=int64)

- model.predict_proba(features_train[250:255]) : 확률값이 나옴. 

In [99]:
model.predict_proba(features_train[250:255])

array([[0.3873625 , 0.6126375 ],
       [0.29956955, 0.70043045],
       [0.12736921, 0.87263079],
       [0.21876764, 0.78123236],
       [0.5817178 , 0.4182822 ]])

##### 평가
- ** 실무에선 사용 안함. 
- from sklearn.metrics import accuracy_score 

In [100]:
target_train_predict = model.predict(features_train)
target_train_predict.shape # target_train.shape과 동일

(535,)

In [101]:
from sklearn.metrics import accuracy_score # 정확도를 평가함

In [102]:
accuracy_score(target_train, target_train_predict) # 내부 평가
# 정규화 이전: 0.7065420560747664

0.708411214953271

In [103]:
target_test_predict = model.predict(features_test)
target_test_predict.shape # target_test.shape 동일

(179,)

In [104]:
accuracy_score(target_test, target_test_predict) # 외부 평가
# 정규화 이전: 0.659217877094972

0.6480446927374302

- accuracy_score(target_train, target_train_predict) # 내부 평가 _ 0.7065420560747664 
- accuracy_score(target_test, target_test_predict) # 외부 평가 _ 0.659217877094972

##### 정규화 이후 
- accuracy_score(target_train, target_train_predict) 내부 평가: 0.708411214953271
- accuracy_score(target_test, target_test_predict) 외부평가: 0.6480446927374302

#### 평가 주요 사용 
- from sklearn.metrics import classification_report _ 종합적으로 나옴
    - precision, recall, f1-score, support : 이 중에 f1-score를 중요하게 볼 것. (판단에 필요한 score) 
    - survived의 0, 1의 일치하는 score임_각 항목중 f1-score를 보고 판단함.  
    - macro avg를 볼 것. : macro avg는 0,1 의 각 항목 평균값을 나타냄. precision과 recall과 f1-score은 macro avg를 주요하게 봄.
    

기존 _ train
  precision    recall  f1-score   support

           0       0.71      0.83      0.77       312
           1       0.69      0.54      0.60       223

    accuracy                           0.71       535
   - macro avg       0.70      0.68      0.69       535
- weighted avg       0.70      0.71      0.70       535
기존 _ test
precision    recall  f1-score   support

           0       0.73      0.72      0.73       112
           1       0.54      0.55      0.55        67

    accuracy                           0.66       179
   - macro avg       0.64      0.64      0.64       179
- weighted avg       0.66      0.66      0.66       179

In [105]:
from sklearn.metrics import classification_report

In [106]:
print (classification_report(target_train, target_train_predict))

              precision    recall  f1-score   support

           0       0.72      0.82      0.77       312
           1       0.69      0.55      0.61       223

    accuracy                           0.71       535
   macro avg       0.70      0.69      0.69       535
weighted avg       0.71      0.71      0.70       535



In [107]:
print (classification_report(target_test, target_test_predict))

              precision    recall  f1-score   support

           0       0.72      0.71      0.71       112
           1       0.53      0.55      0.54        67

    accuracy                           0.65       179
   macro avg       0.63      0.63      0.63       179
weighted avg       0.65      0.65      0.65       179



test예측값은 train 값보다 f1-score의 0과 1 사이가 조금 차이남. 

#### 오차행렬

In [33]:
from sklearn.metrics import confusion_matrix , precision_score, recall_score, f1_score

In [34]:
confusion_matrix(target_train, target_train_predict)

array([[258,  54],
       [103, 120]], dtype=int64)

In [35]:
precision_score(target_train, target_train_predict)

0.6896551724137931

In [36]:
recall_score(target_train, target_train_predict)

0.5381165919282511

In [37]:
f1_score(target_train, target_train_predict)

0.6045340050377833